### DISCUSSION TITLE
Viomi S9 robot vacuum integration

### DISCUSSION THREAD

**Discussion by bfg100k** [PiotrMachowski/lovelace-xiaomi-vacuum-map-card#701] [Show and tell]
So I’ve got a Viomi S9 robot vacuum and while I’m happy with its own app interface and notifications, I figured I can get even more out of it via integration with HA. After a couple of months of playing around, here’s what I’ve managed to achieve.

1.Ditch the mihome app as I have (almost) complete control and visibility of the vacuum via HA mobile app.
a) mobile vacuum dashboard
<img src="https://github.com/PiotrMachowski/lovelace-xiaomi-vacuum-map-card/assets/2188087/fa4d4f0e-98a7-4e50-adb5-3375971cb20d" width="230" height="500"><img src="https://github.com/PiotrMachowski/lovelace-xiaomi-vacuum-map-card/assets/2188087/0173ca50-a1d4-4734-86a3-724fb4f7f31b" width="230" height="500">

b) Changing clean type to rooms, zone or point
<img src="https://github.com/PiotrMachowski/lovelace-xiaomi-vacuum-map-card/assets/2188087/f66b7082-c88a-4e91-bd86-4e9f42fefe3a" width="230" height="500"><img src="https://github.com/PiotrMachowski/lovelace-xiaomi-vacuum-map-card/assets/2188087/c27e2049-3c26-4c4a-9191-7952e926c317" width="230" height="500">

c) Changing clean mode to vacuum, vacuum & mop or just mop
<img src="https://github.com/PiotrMachowski/lovelace-xiaomi-vacuum-map-card/assets/2188087/4460d1f3-6317-4c81-8dd5-98964fe9c477" width="230" height="500">

d) Changing vacuum and mop options (relevant options are shown depending on clean mode)
<img src="https://github.com/PiotrMachowski/lovelace-xiaomi-vacuum-map-card/assets/2188087/75712323-7350-4170-bfbd-d48a6d1b43c3" width="230" height="500"><img src="https://github.com/PiotrMachowski/lovelace-xiaomi-vacuum-map-card/assets/2188087/4edb9836-1633-4164-a4ad-fe7ca7621649" width="230" height="500"><img src="https://github.com/PiotrMachowski/lovelace-xiaomi-vacuum-map-card/assets/2188087/3c5438fd-6186-4122-a565-2266435e4d55" width="230" height="500"><img src="https://github.com/PiotrMachowski/lovelace-xiaomi-vacuum-map-card/assets/2188087/10809270-c913-45c3-acef-2939e30bb7c3" width="230" height="500">

2. Advanced mobile (and watch!) notifications
<img src="https://github.com/PiotrMachowski/lovelace-xiaomi-vacuum-map-card/assets/2188087/a6690133-b46e-431f-a50b-cb81ccb3833a" width="230" height="500">

a) Summary of stats upon job completion including map screenshot (so I can see if the vac missed any area)
<img src="https://github.com/PiotrMachowski/lovelace-xiaomi-vacuum-map-card/assets/2188087/90b53727-fceb-4671-a255-a5af160dbc41" width="230" height="500"><img src="https://github.com/PiotrMachowski/lovelace-xiaomi-vacuum-map-card/assets/2188087/4d249dfc-a626-4dd8-8bfd-9d75e4cac3c7" width="230" height="500">

b) Daily vacuum reminder - sent at 1pm only if I haven’t run the vacuum in the morning (bonus - I can start the cleaning right from the notification!)
<img src="https://github.com/PiotrMachowski/lovelace-xiaomi-vacuum-map-card/assets/2188087/229c008f-75fa-411d-8382-80407b0aa0c4" width="30%">

3. Given that the HA dasboard is accessible via web, I now have full control and visibility of the vacuum on my desktop (unlike the native Viomi interface which is app only)
<img src="https://github.com/PiotrMachowski/lovelace-xiaomi-vacuum-map-card/assets/2188087/5afe2aee-c518-4298-97d7-f1408dc19b66" width="100%">

4. I own a Google Nest Hub which is also integrated with HA so I can now access the same UI on it too instead of the default basic vacuum UI. I have also created an automation helper that will cast the vacuum dashboard to the Nest Hub (if its not playing other media) every time a cleaning job starts so I can monitor the cleaning process.
<img src="https://github.com/PiotrMachowski/lovelace-xiaomi-vacuum-map-card/assets/2188087/ce9cc226-0557-4c80-be3f-04dd69ca236d" width="60%">

5. Speaking of automation, I also have an automation helper that will change the clean mode (i.e. vacuum or mop) automatically when the mop pad is physically installed/removed. IMHO, this is the most useful feature out of all the above as I can simply install/remove the pad and say “Hey Google, start the vacuum” knowing that the right job will run! (Before this, I almost always forget to change the mode in the app before hitting the start button)

**Comment by PiotrMachowski** [comment]
Looks great! Could you share your configs?

**Reply by bfg100k** [reply]
Yep sure. Have uploaded my config and scripts [here](https://github.com/bfg100k/Viomi-S9-HA-Integration/tree/main)

**Comment by YOYONL89** [comment]
Hi bfg100k,

I was looking at your scripts and try to modifi them to use with a roborock s1 and i have run a snag.
In your comment of the Viomi S9 0 job complete and device error notifications you say the following: `V2 uses the MihomeMessageSensor entity for triggering notifications
  which appears to be quicker and more accurate.`

i cant find that specific entity nor i cant find in HACS.
where did you get this? because i can see you use it in your automation i think: `sensor.mi_1838149435_message` is the entity?

**Reply by bfg100k** [reply]
This sensor is automatically created by the Xiaomi MIOT Auto integration. Which backend integration are you using?

**Reply by YOYONL89** [reply]
I use the same backend components as you do.
In the Xiaomi MIOT Auto integration i have even enabled the cloud option.

It could be that the vacuum is to old/has different firmware which does not expose it.

The entity with the most info is: 
`vacuum.roborock_m1s_1efc_robot_cleaner`

```
Attributen
Model
roborock.vacuum.m1s
Lan IP
Something
MAC address
Something
Entity class
MiotRoborockVacuumEntity
Miot type
urn:miot-spec-v2:device:vacuum:0000A006:roborock-m1s:2
Room mapping
- - 16
- '758001110737'
- Woonkamer
- - 17
- '758001110727'
- Keuken
Vacuum.mode
3
Vacuum.status
8
State updater
lan
Sub entities
segment_16, segment_17, vacuum-2.status-3, battery-4.battery_level-1, props:clean_area, props:clean_time
Vacuum.status desc
Charging
Props
msg_ver: 1
msg_seq: 932
state: 8
battery: 100
clean_time: 18
clean_area: 0
error_code: 0
map_present: 1
in_cleaning: 0
in_returning: 0
in_fresh_state: 1
lab_status: 1
fan_power: 103
dnd_enabled: 0
map_status: 3
lock_status: 0
main_brush_work_time: 195689
side_brush_work_time: 195689
filter_work_time: 195689
sensor_dirty_time: 62888
fan_mode: 103
Clean area
0
Clean time
0,3
Miot action result
code: 0
siid: 2
aiid: 1
result:
- ok

```

As you can see I can get the error codes out but the message attribute is not present.
Maybe I just have to get something around this

**Reply by bfg100k** [reply]
I believe this sensor is tied to your xiaomi account and not your device so you should have it as long as the integration is setup correctly. FYI I have a couple of xiaomi devices connected to HA but only have 1 message sensor.
